### ISSUE TITLE
Assertion failure on linux

### ISSUE THREAD

**Issue by jeremysalwen** [quicky2000/tartini#28]
I a getting an assertion failure when running in debug mode on linux:

Assert Failed: l_current_time + m_loop_step < (int)m_channel->get_pitch_lookup_smoothed().size() in ../tartini/sound/notedata.cpp line 117.

I added some debug printing, here is the log:

lookup=1
lookup=2
lookup=3
lookup=4
lookup=5
lookup=6
lookup=7
lookup=8
lookup=9
lookup=10
lookup=11
lookup=12
lookup=13
lookup=14
lookup=15
lookup=16
lookup=17
lookup=18
lookup=19
lookup=20
lookup=21
lookup=22
lookup=23
lookup=24
lookup=25
lookup=26
lookup=27
lookup=28
lookup=29
lookup=30
lookup=31
lookup=32
lookup=33
lookup=34
lookup=35
lookup=36
lookup=37
lookup=38
lookup=39
lookup=40
lookup=41
lookup=42
lookup=43
lookup=44
lookup=45
lookup=46
lookup=47
NoteData Constructor m_loop_start=1516 m_start_chunk=46 framesPerChunk()=32 m_loop_step=44
lookup=48
lookup=49
lookup=50
lookup=51
Before Assertion m_loop_start=1516 current_time=1516 loop_step=44 pitch_lookup_size=816
Assert Failed: l_current_time + m_loop_step < (int)m_channel->get_pitch_lookup_smoothed().size() in ../tartini/sound/notedata.cpp line 117.

(lookup= is an existing printf from channel.cpp)

This error doesn't occur in the original tartini.  I added similar printf statements, and I got 

NoteData Constructor m_loop_start=2092 m_start_chunk=2 framesPerChunk()=32 loop_step=44
Before Assertion m_loop_start=2092 current_time=2092 loop_step=44 pitch_lookup_size=3072

Do you have any idea where you might have introduced this bug?

**Comment by quicky2000** [comment]
Hello @jeremysalwen 

In fact if I remember well when we tried to use original Tartini on Linux we sometimes faced some crash.
By investigating around that we found that the reason was an out of bound access that was caused by **l_current_time + m_loop_step** greater than size of vector **pitch_lookup_smoothed**.
At this time the bug is still not solved but we add an assert to detect such out of bound access in order to debug it.
This is exactly what occur for you.
With original tartini the check is not there so out of bounds access are performed and if you are lucky there is no crash but the bug is there.
We didn't introduced the bug, we added a systematic symptoma of the bug but for sure we need to understand the bug in order to correct it
Is it more clear ?

**Comment by jeremysalwen** [comment]
Hmm, I think this is a different check than the one you are thinking of.  The check was present in the original tartini source: https://github.com/jeremysalwen/tartini-debian/blob/master/sound/notedata.cpp#L124 

You can also see from the debug data I output above that the constraint is satisfied in the original version.

**Comment by quicky2000** [comment]
yes you are right, i was confused with some other asserts we add.
Normally I didn't modify Tartini algorithm, perhaps I missed something.
How do you proceed to use the same data on both old version and mine one ?
you have an audio file that you could share so I can try to locate the issue ?

**Comment by jeremysalwen** [comment]
I see it almost immediately when recording any audio.  So the data is not exactly the same, but it's highly reproducible.

**Comment by quicky2000** [comment]
ok. on which commit are you exactly ? the location of assert is not the one I observe in my latest commit.

**Comment by jeremysalwen** [comment]
I think the only difference is some added print (moving it down a couple lines).  I uploaded my exact version to https://github.com/jeremysalwen/tartini

**Comment by quicky2000** [comment]
Hello. Could you give more details on your configuration: Qt, Qwt versions ? compiler used etc ?
On my side with g++ 7.4.0 , Qt 5.9.5, Qwt 6.1.3 on Kubuntu 18.04 I tested several times and I don't succeed to reproduce the issue

**Comment by jeremysalwen** [comment]
g++ (Debian 8.3.0-6) 8.3.0
qt 5.11.3
qwt 6.1.4-1
on a debian-derived distro.

I think it may be hard for you to reproduce, but if you have any ideas for what the root cause of this change was, I could run any instrucmentation, or debug myself.

**Comment by quicky2000** [comment]
What you could try is comment line 29 of myassert.h so that condition failure still print the message but don't stop execution.
Run Tartini with its recording function and stop it after several assertion failure display.
Then try to open the resulting wave file with tartini and check if assertion failure message are printed.
if yes you could share the wave file with me and we would be able to compare executions

**Comment by quicky2000** [comment]
Problem also encountered by @mself on Mac OS.
The good news it that he obtains it with [example.wav](https://miracle.otago.ac.nz/tartini/download/wav/example.wav) from Tartini website so it will allow to compare with my own execution that don't crash

I copy/paste the report from @mself

> NoteData::addVibratoData:

> 114: myassert(l_current_time + m_loop_step < (int)m_channel->get_pitch_lookup_smoothed().size());

> lldb) p l_current_time
> (int) $3 = 1388
> (lldb) p m_loop_step
> (int) $4 = 44
> (lldb) p m_channel->get_pitch_lookup_smoothed().size()
> (uint) $5 = 720

Backtrace:

>   * frame #3: 0x00000001000ba84e tartini`NoteData::addVibratoData(this=0x00000001107102f8, p_chunk=44) at notedata.cpp:114:13
>     frame #4: 0x0000000100015877 tartini`Channel::calcVibratoData(this=0x00000001038b54d0, p_chunk=44) at channel.cpp:1197:25
>     frame #5: 0x0000000100091f7b tartini`MyTransforms::calculateAnalysisData(this=0x00000001038913d0, p_chunk=44, p_channel=0x00000001038b54d0) at mytransforms.cpp:564:24
>     frame #6: 0x00000001000106cc tartini`SoundFile::calculateAnalysisData(this=0x00000001038913a0, p_chunk=44, p_channel=0x00000001038b54d0) at soundfile.hpp:186:21
>     frame #7: 0x00000001000104e7 tartini`Channel::processNewChunk(this=0x00000001038b54d0, p_filter_state=0x00007ffeefbfca60) at channel.cpp:325:15
>     frame #8: 0x000000010002b0e5 tartini`SoundFile::readChunk(this=0x00000001038913a0, p_n=32) at soundfile.cpp:529:26
>     frame #9: 0x000000010002b722 tartini`SoundFile::preProcess(this=0x00000001038913a0) at soundfile.cpp:662:11
> 

This crash only happens when I have "Do detailed pitch analysis" selected in preferences.

**Comment by quicky2000** [comment]
I created a branch **issue_28** to investigated and add a debug message.
@mself could you compile it, open example.wav and attach the log you obtain ?
Here is mine
[ok.log.gz](https://github.com/quicky2000/tartini/files/4521706/ok.log.gz)

**Comment by mself** [comment]
For me, it stops on the first iteration:
```
QT_VERSION_STR="5.14.1"
QT_VERSION=50e01
QWT_VERSION_STR="6.1.4"
QWT_VERSION=60104
2020-04-23 09:23:01.792951-0700 tartini[43656:510434] [qt.qpa.fonts] Populating font family aliases took 456 ms. Replace uses of missing font family "AnyStyle" with one that exists to avoid this cost.
Opening file: /Users/Matthew/Projects/Whistle/Sample Whistles/example.wav
(FilteredFilename: /Users/Matthew/Projects/Whistle/Tartini/tartini-master/Debug/temp017.wav)
File: channels=1, freq=44100, bits=16
channels = 1
channel size = 32
----------Opening------------
filename = /Users/Matthew/Projects/Whistle/Sample Whistles/example.wav
rate = 44100
channels = 1
bits = 16
windowSize = 32
stepSize = 32
-----------------------------
current time: 1388	 loop_step: 44	 smoothed size: 720
Assert Failed: l_current_time + m_loop_step < (int)m_channel->get_pitch_lookup_smoothed().size() in /Users/Matthew/Projects/Whistle/Tartini/tartini/sound/notedata.cpp line 115.
Causing a trace crash.
```

**Comment by mself** [comment]
I need to trace back and see where `windowSize` and `stepSize` are getting set to those values.

**Comment by mself** [comment]
I notice that in the log that your `windowSize` and `stepSize` are much larger.  I am using the default values in the preferences panel of 1ms buffer size and 1ms step size.  It looks like your `windowSize` and `stepSize` values correspond to buffer and step size values of 48ms and 24ms, which are the "Cello" settings.

**Comment by mself** [comment]
I noticed that in `GData::getAnalysisBufferSize()` and `GData::getAnalysisStepSize()`, the default values passed to `m_settings->value()` are 48ms and 24ms.  These correspond to `windowSize = 2048` and `stepSize = 1024`, which are the values in your logs.

But when I run the app, the values of 1ms (which is what is shown in the grayed out UI) are used.

So, I'm wondering if the port to Qt5 has changed the behavior of how these default values are being set?  I'm not experience with using Qt, but perhaps the default values need to also be set for the widgets in the designer?  Or maybe these widgets used to be active (not grayed out) before?

If this is the cause, then there still is a bug in the detailed pitch analysis -- but it only occurs for certain small values of the buffer size and step size settings (and not for the default values).

**Comment by mself** [comment]
There are definitely a few things wrong with the Settings dialog.

When you first launch the app and open Settings, the Note Range is set to Custom and the default values of Buffer/Step Size in the Analysis tab are 1ms, which are bad values.  If you change those values, they don't persist if you close and reopen the Settings dialog.

If you select one of the Note Range options from the dropdown in the General tab, it sets the Buffer/Step Size groups in the Analysis tab to specific values (and also disables those control groups).  This works as expected.  But if you close and reopen the Settings dialog, the Buffer/Step Size will again get reset to 1ms even though the Note Range drop down retains its previous value.

Also, if you press the Restore Defaults button, it sets the Note Range to "Cello", but also resets the Buffer/Step Size in the Analysis tab are 1ms.

Basically, every time you open the Settings dialog it sets the Buffer/Step Size to 1ms and unless you manually pick a value from the Note Range dropdown you will end up in a bad state.

**Comment by mself** [comment]
I found the bug.  It was treating those fields as `bool` rather than `int`:
```
    else if(l_class_name == "QSpinBox")
    {
        ((QSpinBox*)p_object)->setValue(g_data->getSettingsBoolValue(l_full_key));
    }
```
needs to be:
```
    else if(l_class_name == "QSpinBox")
    {
        ((QSpinBox*)p_object)->setValue(g_data->getSettingsIntValue(l_full_key));
    }
```

**Comment by quicky2000** [comment]
I confirm in my config I have
`bufferSizeUnit=milli-seconds`
`bufferSizeValue=48`
`stepSizeUnit=milli-seconds`
`stepSizeValue=24`
and Note Ranges is on
`Notes F1 and higher -  - Cello, Guitar, Bass Clarinet, General sounds ...`

On my side I have an issue in save configuration file with 
`[General]`
that is corrupted to
`[%General]`

**Comment by mself** [comment]
With this change the Buffer/Step Size values get set appropriately and the app no longer crashes when loading files.

Note that there still is a bug in that if you do set these values to 1ms/1ms the app does crash.  There may need to be some additional logic on what values are supported (e.g. maybe the Step size needs to be smaller than the Buffer size?).  But this issue seems much less urgent.

**Comment by quicky2000** [comment]
Well done !
Thanks for your analysis I found that it was introduced in commit dc983c41b2a2565e7405685dba2fc52a1159284f
It was due to a bad copy/paste :-(
Could you do a Pull Request of your modification ? by this way it will be credited to you

**Comment by quicky2000** [comment]
> Note that there still is a bug in that if you do set these values to 1ms/1ms the app does crash. There may need to be some additional logic on what values are supported (e.g. maybe the Step size needs to be smaller than the Buffer size?). But this issue seems much less urgent.

Indeed I have the same crash if I use 1ms for both parameters but it is OK if buffer size is 2ms.
I will look in the code to understand why and then add a check to settings dialog to avoid this

**Comment by quicky2000** [comment]
Hi @jeremysalwen could you check if the patch solve the fissue for you too ? It is now integrated in master branch
